### PR TITLE
feat(infra): doc properties by orm

### PR DESCRIPTION
**Pull request by EYHN**
create new orm table docConfiguration

move primary store to docConfiguration

### CODE REVIEW

**Comment by EYHN** [comment]
* **#8382** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/8382?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @EYHN and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

**Review by pengx17 — APPROVED** [review]

File: packages/common/infra/src/utils/yjs-observable.ts [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+}
+
+/**
+ * convert yjs type to observable.
+ * observable will automatically update when yjs data changed.
+ *
+ * @example
+ * yjsObserveDeep(yjs) -> emit when any of its deep children changed
+ */
+export function yjsObserveDeep(yjs?: any) {
```
Comment (pengx17) [inline packages/common/infra/src/utils/yjs-observable.ts:84]: @Saul-Mirone can you help review this function?

File: packages/common/infra/src/modules/doc/stores/doc-properties.ts [hunk 2]
```diff
@@ -0,0 +1,229 @@
...
+          latency?.type ??
+          'unknown' /* should never reach here, just for safety, we need handle unknown property type */,
+        name: latency?.name,
+        ...config,
+      });
+      return;
+    } else {
+      this.dbService.db.docCustomPropertyInfo.update(id, config);
+    }
+  }
```
Comment (pengx17) [inline packages/common/infra/src/modules/doc/stores/doc-properties.ts]: may split the migration code for clearity

```suggestion
updateDocPropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
  const needMigration = !this.dbService.db.docCustomPropertyInfo.get(id);
  return needMigration
    ? this.migrateAndCreatePropertyInfo(id, config)
    : this.dbService.db.docCustomPropertyInfo.update(id, config);
}

private migrateAndCreatePropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
  const latencyId = id.replace('custom:', '');
  const latency = this.getLatencyDocPropertyInfo(latencyId);
  return this.dbService.db.docCustomPropertyInfo.create({
    id,
    type: latency?.type ?? 'unknown',
    name: latency?.name,
    ...config,
  });
}
```

File: packages/common/infra/src/modules/doc/stores/doc-properties.ts [hunk 3]
```diff
@@ -0,0 +1,229 @@
...
+      additionalData: {}, // also remove additional data to reduce size
+      isDeleted: true,
+    });
+  }
+
+  updateDocPropertyInfo(id: string, config: Partial<DocCustomPropertyInfo>) {
+    const needMigration = !this.dbService.db.docCustomPropertyInfo.get(id);
+    if (needMigration) {
+      // if this property is not in db, we need to migration it from latency to db, only type and name is needed
+      const latencyId = id.replace('custom:', '');
```
Comment (pengx17) [inline packages/common/infra/src/modules/doc/stores/doc-properties.ts:74]: do you mean "legacy"?

File: packages/common/infra/src/modules/doc/stores/doc-properties.ts [hunk 4]
```diff
@@ -0,0 +1,229 @@
...
+  yjsObserveDeep,
+} from '../../../utils/yjs-observable';
+import type { WorkspaceDBService } from '../../db';
+import type {
+  DocCustomPropertyInfo,
+  DocProperties,
+} from '../../db/schema/schema';
+import type { WorkspaceService } from '../../workspace';
+
+interface LatencyDocProperties {
```
Comment (pengx17) [inline packages/common/infra/src/modules/doc/stores/doc-properties.ts]: Do you mean `LegacyDocProperties`?

File: packages/common/infra/src/utils/yjs-observable.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+      };
+      refresh();
+      target.observe(refresh);
+      return () => {
+        target.unobserve(refresh);
+      };
+    }).pipe(
+      distinctUntilChanged(),
+      switchMap(arr => _yjsDeepWatch(arr, path.slice(1)))
+    );
```
Comment (pengx17) [inline packages/common/infra/src/utils/yjs-observable.ts]: the two branches are almost the same. perhaps merge them?

File: packages/common/infra/src/utils/yjs-observable.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  if (target instanceof YArray) {
+    return new Observable(subscriber => {
+      const refresh = () => {
+        if ('index' in current) {
+          subscriber.next(target.get(current.index));
+        } else {
+          subscriber.next(undefined);
+        }
+      };
+      target.observe(refresh);
```
Comment (pengx17) [inline packages/common/infra/src/utils/yjs-observable.ts]: `refresh` on init?

File: packages/common/infra/src/utils/yjs-observable.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+ * this function is optimized for deep watch performance.
+ *
+ * @example
+ * yjsObserveByPath(yjs, 'pages.[0].id') -> only emit when pages[0].id changed
+ * yjsObserveByPath(yjs, 'pages.[0]').switchMap(yjsObserve) -> emit when any of pages[0] or its children changed
+ * yjsObserveByPath(yjs, 'pages.[0]').switchMap(yjsObserveDeep) -> emit when pages[0] or any of its deep children changed
+ */
+export function yjsObserveByPath(yjs: YAbstractType<any>, path: string) {
+  const parsedPath = parsePath(path);
+  return _yjsDeepWatch(yjs, parsedPath);
```
Comment (pengx17) [inline packages/common/infra/src/utils/yjs-observable.ts:74]: What if the target being watched hasn't been created yet at the time of observation? Should we also observe its parent in this case?

Comment (EYHN) [reply packages/common/infra/src/utils/yjs-observable.ts:74]: We will observe the parent until the child is created

**Review by CatsJuice — APPROVED** [review]

**Pull request merged**